src/middlewared/middlewared/utils/account/utmp.py

```diff
@@ -0,0 +1,245 @@
+
+
+def __setutent():
+    # Rewind the file pointer to beginning of utmp file
+    # void setutent(void);
+    func = libc.setutent
+    func.argtypes = []
+    func.restype = None
+    func()
+
+
+def __endutent():
+    # Close the utmp file
+    # void endutent(void);
+    func = libc.endutent
+    func.argtypes = []
+    func.restype = None
+    func()
+
+
+def __utmpname(file: LoginFile):
+    # Sets the name of the utmp-format file for other utmp file names to access
+    func = libc.utmpname
+    func.argtypes = [ctypes.c_char_p]
+    func.restype = ctypes.c_int
+
+    res = int(func(ctypes.c_char_p(str(file).encode())))
+    if res != 0:
+        raise RuntimeError(f'utmpname() failed with error: {ctypes.get_errno()}')
+
+
+def __getutent():
+    # Read a line from the current position in the utmp file. We don't bother with
+    # re-entrant versions because we're taking a global lock due to concerns about file position
+    # struct utmp *getutent(void);
+    func = libc.getutent
+    func.argtypes = []
+    func.restype = ctypes.c_void_p
+
+    res = func()
+    return ctypes.cast(res, ctypes.POINTER(StructUtmp))
+
+
+def __parse_utmp_exit(utmp_type: PyUtmpType, data: StructExitStatus) -> PyUtmpExit:
+    if utmp_type != PyUtmpType.DEAD_PROCESS:
+        return None
+
+    return PyUtmpExit(e_termination=data.e_termination, e_exit=data.e_exit)
+
+
+def __parse_timeval(data: StructTimeval) -> datetime:
+    secs = data.tv_sec
+    secs += (data.tv_usec / 1000000)
+    if secs == 0:
+        return None
+
+    return datetime.fromtimestamp(secs, UTC)
+
+
+def __parse_address(int_array) -> IPv4Address | IPv6Address | None:
+    if not int_array[0]:
+        return None
+
+    if not int_array[1]:
+        # IPv4 Address only uses ut_addr_v6[0]
+        try:
+            return IPv4Address(ntohl(int_array[0]))
+        except Exception:
+            return None
+
+    ipv6_val = ntohl(int_array[0])
+    ipv6_val += ntohl(int_array[1]) << 32
+    ipv6_val += ntohl(int_array[2]) << 64
+    ipv6_val += ntohl(int_array[3]) << 96
+
+    try:
+        return IPv6Address(ipv6_val)
+    except Exception:
+        return None
+
+
+def __parse_utmp_entry(entry: StructUtmp) -> PyUtmpEntry:
+    # Convert our utmp buffer to python dataclass
+    return PyUtmpEntry(
+        ut_type=PyUtmpType(entry.contents.ut_type),
+        ut_pid=entry.contents.ut_pid or None,
+        ut_line=entry.contents.ut_line.decode() or None,
+        ut_id=entry.contents.ut_id.decode(),
+        ut_user=entry.contents.ut_user.decode() or None,
+        ut_host=entry.contents.ut_host.decode() or None,
+        ut_exit=__parse_utmp_exit(entry.contents.ut_type, entry.contents.ut_exit),
+        ut_session=entry.contents.ut_session,
+        ut_tv=__parse_timeval(entry.contents.ut_tv),
+        ut_addr=__parse_address(entry.contents.ut_addr)
+    )
+
+
+def __extend_utmp_entry(entry: StructUtmp) -> dict:
+    data = asdict(entry)
```

Comment:
Not sure if this is a hot path but asdict does a deepcopy fyi.

Reply:
It's probably not too bad compared to other fileio we're doing here.